### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by jmprieur**
- Upgrading the sample to ADAL.NET 3.19.2
- Moving the called API from AAD Graph to Microsoft Graph, show-casing a REST call (hence changing the appid)
- Adding app creation scripts and doc
- Fixing code issues / code smells / static analysis issues.
- Updating the README.md, including metadata, links to docs, and topology

### CODE REVIEW

File: DirSearcherClient/Program.cs [hunk 1]
```diff
@@ -75,13 +73,9 @@ static void Search(string searchterm, string tenant)
...
             {
                 JObject jResult = null;
-
-                string graphResourceUri = "https://graph.windows.net";
-                string graphApiVersion = "2013-11-08";
-
                 try
                 {
-                    string graphRequest = String.Format(CultureInfo.InvariantCulture, "{0}/{1}/users?api-version={2}&$filter=mailNickname eq '{3}'", graphResourceUri, ar.TenantId, graphApiVersion, searchterm);
+                    string graphRequest = $"https://graph.microsoft.com/v1.0/users?$filter=mailNickname eq '{searchterm}'";
```
Comment (henrik-me) [inline DirSearcherClient/Program.cs]: >https://graph.microsoft.com [](start = 44, length = 27)

{resource} ?

File: DirSearcherClient/Program.cs [hunk 2]
```diff
@@ -97,27 +91,30 @@ static void Search(string searchterm, string tenant)
...
 
-                if (jResult["odata.error"] != null || jResult["value"] == null)
+                if (jResult == null || jResult["odata.error"] != null || jResult["value"] == null)
                 {
                     Console.ForegroundColor = ConsoleColor.Red;
                     Console.WriteLine("Error on search");
                 }
                 Console.ForegroundColor = ConsoleColor.Green;
-                if (jResult.Count == 0)
+                if (jResult != null)
```
Comment (henrik-me) [inline DirSearcherClient/Program.cs]: >if (jResult != null) [](start = 16, length = 20)

invert and return or return above where you anyway state we are in an error situation.

File: DirSearcherClient/Program.cs [hunk 3]
```diff
@@ -97,27 +91,30 @@ static void Search(string searchterm, string tenant)
                     Console.WriteLine(ee.Message);
                 }
 
-                if (jResult["odata.error"] != null || jResult["value"] == null)
+                if (jResult == null || jResult["odata.error"] != null || jResult["value"] == null)
                 {
                     Console.ForegroundColor = ConsoleColor.Red;
                     Console.WriteLine("Error on search");
```
Comment (henrik-me) [inline DirSearcherClient/Program.cs]: >Console.WriteLine("Error on search"); [](start = 20, length = 37)

return after this line?

**Comment by henrik-me** [comment]
>                 catch (Exception ee)

specific exception?

---
Refers to: DirSearcherClient/Program.cs:87 in 5a778c6. [](commit_id = 5a778c6c6d2c997db5159c621445d8405806bcdd, deletion_comment = False)

File: DirSearcherClient/Program.cs [hunk 4]
```diff
@@ -97,27 +91,30 @@ static void Search(string searchterm, string tenant)
                     Console.WriteLine(ee.Message);
```
Comment (henrik-me) [inline DirSearcherClient/Program.cs]: >Console.WriteLine(ee.Message); [](start = 20, length = 30)

return after this? does it make sense to do the next stuff?

**Comment by henrik-me** [comment]
>             }

looks like this should be a generic helper function available to all samples... or perhaps as part of the library...

---
Refers to: DirSearcherClient/Program.cs:175 in 5a778c6. [](commit_id = 5a778c6c6d2c997db5159c621445d8405806bcdd, deletion_comment = False)

File: DirSearcherClient/Program.cs [hunk 5]
```diff
@@ -181,23 +178,29 @@ static AuthenticationResult GetToken(string tenant)
             {
                 result = ctx.AcquireTokenSilentAsync(resource, clientId).Result;
             }
+            catch(AggregateException ex) when (ex.InnerException is AdalSilentTokenAcquisitionException)
+            {
+                result = GetTokenViaCode(ctx);
+            }
+            catch (AdalSilentTokenAcquisitionException)
```
Comment (henrik-me) [inline DirSearcherClient/Program.cs]: >(AdalSilentTokenAcquisitionException) [](start = 18, length = 37)

Minor... combine with catch above ? catch(ex) when (ex is ...) || (ex is ...) ? can e.g. be done via an exception filtering method (private bool IsAdalSilentTokenAquisitionException(exception e) {}) ... not sure though if this works for .net core...

File: README.md [hunk 6]
```diff
@@ -1,124 +1,198 @@
...
 
-The ASP.NET documentation pages provide step by step instructions for installing .NET Core (the .NET Execution Environment) for your platform of choice.
-#### Windows
-If you are targeting Windows, see [this document](https://docs.asp.net/en/latest/getting-started/installing-on-windows.html).
-(this readme assumes you will follow the instructions for installing ASP.NET 5 with Visual Studio, but you can also go for the standalone option).
-#### Mac OS X
-If you are targeting Mac OS X, please follow the instructions [here](https://docs.asp.net/en/latest/getting-started/installing-on-mac.html). 
-#### Linux
-If you are targeting Linux, please follow the instructions [here](https://docs.asp.net/en/latest/getting-started/installing-on-linux.html). This sample has been tested on Ubuntu 
+Once you have completed the DNX setup, from your shell or command-line run:
```
Comment (henrik-me) [inline README.md]: >DNX [](start = 28, length = 3)

I think this terminology should be explained. e.g. is that all the pre-reqs above or something else?

**Comment by henrik-me** [comment]
Comment regarding Topology picture; Please see: [http://www.dictionary.com/browse/thru](http://www.dictionary.com/browse/thru) (not sure we should use informal language)

File: DirSearcherClient/Program.cs (hunk 1 above)
Comment (jmprieur) [reply DirSearcherClient/Program.cs]: Thanks. Done

File: DirSearcherClient/Program.cs (hunk 2 above)
Comment (jmprieur) [reply DirSearcherClient/Program.cs]: I returned above. Thanks for catching this.

File: DirSearcherClient/Program.cs (hunk 5 above)
Comment (jmprieur) [reply DirSearcherClient/Program.cs]: yes, I agree.

File: README.md (hunk 6 above)
Comment (jmprieur) [reply README.md]: Good catch.

**Pull request merged**